score_models/analytic_models/conv_likelihood.py

```diff
@@ -0,0 +1,115 @@
+from typing import Callable, Union, Tuple, Optional
+import torch
+from torch import Tensor
+import torch.nn as nn
+from torch import vmap
+import numpy as np
+
+from ..sde import SDE
+
+
+class ConvolvedLikelihood(nn.Module):
+    """
+    Convolved likelihood approximation for the likelihood component of a
+    posterior score model.
+
+    Applies the convolved likelihood approximation as described in Adam et al.
+    2022 Appendix A. Essentially this assumes that the posterior convolution may
+    be approximately factored into the convolution of the likelihood and the
+    prior separately. For a linear and gaussian likelihood model, this is exact,
+    coming out to:
+
+    .. math::
+
+        p_t(y|x_t) = N(y|Ax_t, \Sigma_y + \\sigma_t^2 AA^T)
+
+    We implement this as an energy model, where the energy is the negative log
+    likelihood in the observation space. Autodiff then propagates the score to
+    the model space.
+
+    Args:
+        sde: The SDE that the score model is associated with.
+        y: The observation.
+        Sigma_y: The observation covariance matrix. If 1D this is assumed to be the diagonal of the covariance matrix.
+        x_shape: The shape of the model space.
+        A: The linear operator relating the model space to the observation space.
+        f: The function relating the model space to the observation space.
+        AAT: The covariance of the linear operator.
+        diag: Whether to use the diagonal approximation.
+    """
+
+    @torch.no_grad()
+    def __init__(
+        self,
+        sde: SDE,
+        y: Tensor,
+        Sigma_y: Tensor,
+        x_shape: Tuple[int],
+        A: Optional[Tensor] = None,
```

Comment:
It is common for this approximation to use the functional representation of the forward model. So either add a way to define a forward model or A should accept callable types

Reply:
currently if you want a callable then you use the `f` argument. I bet I can merge them, I'll work on it